pkg/kotsadmsnapshot/backup.go

```diff
@@ -191,254 +201,512 @@ func CreateApplicationBackup(ctx context.Context, a *apptypes.App, isScheduled b
 		}
 		if parentSequence == -1 {
 			// no version is deployed for this app yet
 			continue
 		}
 
-		archiveDir, err := ioutil.TempDir("", "kotsadm")
+		archiveDir, err := os.MkdirTemp("", "kotsadm")
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to create temp dir for app %s", a.Slug)
+			return metadata, errors.Wrapf(err, "failed to create temp dir for app %s", app.Slug)
 		}
-		defer os.RemoveAll(archiveDir)
+		defer func() {
+			_ = os.RemoveAll(archiveDir)
+		}()
 
-		err = store.GetStore().GetAppVersionArchive(a.ID, parentSequence, archiveDir)
+		err = store.GetStore().GetAppVersionArchive(app.ID, parentSequence, archiveDir)
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to get app version archive for app %s", a.Slug)
+			return metadata, errors.Wrapf(err, "failed to get app version archive for app %s", app.Slug)
 		}
 
 		kotsKinds, err := kotsutil.LoadKotsKinds(archiveDir)
 		if err != nil {
-			return nil, errors.Wrap(err, "failed to load kots kinds from path")
+			return metadata, errors.Wrap(err, "failed to load kots kinds from path")
+		}
+
+		metadata.apps[app.Slug] = appInstanceBackupMetadata{
+			app:            app,
+			kotsKinds:      kotsKinds,
+			parentSequence: parentSequence,
+		}
+
+		// if there's only one app, use the slug as the backup name
+		if len(apps) == 1 && len(metadata.apps) == 1 {
+			metadata.backupName = fmt.Sprintf("%s-%d", app.Slug, now.UnixNano())
+		}
+
+		// optimization as we no longer need the archive dir
+		_ = os.RemoveAll(archiveDir)
+	}
+
+	metadata.ec, err = getECInstanceBackupMetadata(ctx, ctrlClient)
+	if err != nil {
+		return metadata, errors.Wrap(err, "failed to get embedded cluster metadata")
+	}
+
+	return metadata, nil
+}
+
+// getECInstanceBackupMetadata returns metadata about the embedded cluster for use in creating an
+// instance backup.
+func getECInstanceBackupMetadata(ctx context.Context, ctrlClient ctrlclient.Client) (*ecInstanceBackupMetadata, error) {
+	if !util.IsEmbeddedCluster() {
+		return nil, nil
+	}
+
+	installation, err := embeddedcluster.GetCurrentInstallation(ctx, ctrlClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current installation: %w", err)
+	}
+
+	seaweedFSS3ServiceIP, err := embeddedcluster.GetSeaweedFSS3ServiceIP(ctx, ctrlClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get seaweedfs s3 service ip: %w", err)
+	}
+
+	return &ecInstanceBackupMetadata{
+		installation:         *installation,
+		seaweedFSS3ServiceIP: seaweedFSS3ServiceIP,
+	}, nil
+}
+
+// getInfrastructureInstanceBackupSpec returns the velero backup spec for the instance backup. This
+// is either the kotsadm backup or the combined backup if this is not using improved DR.
+func getInfrastructureInstanceBackupSpec(ctx context.Context, k8sClient kubernetes.Interface, metadata instanceBackupMetadata, hasAppSpec bool) (*velerov1.Backup, error) {
```

Comment:
IMO, this function name would be clearer without `Instance`
```suggestion
func getInfrastructureBackupSpec(ctx context.Context, k8sClient kubernetes.Interface, metadata instanceBackupMetadata, hasAppSpec bool) (*velerov1.Backup, error) {
```

Reply:
actually, nevermind. it can then be confused as a separate backup type.